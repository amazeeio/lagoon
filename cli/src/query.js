// @flow

import path from 'path';
import os from 'os';
import url from 'url';
import R from 'ramda';
import { config } from './config/';
import { fileExists, readFile } from './util/fs';
import request from './util/request';
import { printErrors } from './printErrors';

type QLQueryArgs = {
  cerr: typeof console.error,
  query: string,
  variables?: Object,
  headers?: Object,
  pretty?: boolean,
};

export async function runGQLQuery({
  cerr,
  query,
  variables,
  headers: customHeaders = {},
  pretty = false,
}:
QLQueryArgs): Object {
  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...customHeaders,
  };

  if (!headers.Authorization) {
    const tokenFile = path.join(os.homedir(), '.lagoon-token');
    const tokenFileExists = await fileExists(tokenFile);

    if (tokenFileExists) {
      const tokenBuffer = await readFile(tokenFile);
      const token = tokenBuffer.toString().replace(/(\r\n|\n|\r)/gm, '');
      headers.Authorization = `Bearer ${token}`;
    }
  }

  const apiUrl = R.prop('api', config) || 'https://api.amazee.io/graphql';

  const {
    hostname, path: pathname, port: urlPort, protocol,
  } = url.parse(
    apiUrl,
  );

  if (hostname == null) {
    throw new Error(
      'API URL configured under the "api" key in .lagoon.yml doesn\'t contain a valid hostname.',
    );
  }

  const body = JSON.stringify(
    {
      query,
      variables,
    },
    null,
    pretty ? 2 : 0,
  );

  const protocolPorts = {
    'https:': 443,
    'http:': 80,
  };

  const options = {
    hostname,
    path: pathname,
    port:
      urlPort === null
        ? R.propOr(443, protocol)(protocolPorts)
        : Number(urlPort),
    method: 'POST',
    headers,
    body,
    rejectUnauthorized: false,
  };

  try {
    return await request(options);
  } catch (err) {
    const error = R.ifElse(
      // For socket hang ups...
      R.propEq('message', 'socket hang up'),
      // ...print a nicer error message...
      R.always('Could not connect to API.'),
      // ...otherwise just return the error message
      R.prop('message'),
    )(err);
    printErrors(cerr, error);
    process.exit(1);
  }
}

export default runGQLQuery;
