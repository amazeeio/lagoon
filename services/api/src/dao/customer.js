const R = require('ramda');
const {
  ifNotAdmin,
  inClause,
  inClauseOr,
  isPatchEmpty,
  knex,
  prepare,
  query,
  whereAnd,
} = require('./utils');

const Sql = {
  updateCustomer: (cred, input) => {
    const { id, patch } = input;
    const { customers } = cred.permissions;

    const updateCustomerQuery = knex('customer').where('id', '=', id);

    if (cred.role !== 'admin' && !R.contains(id, customers)) {
      updateCustomerQuery.whereIn('id', customers);
    }

    return updateCustomerQuery.update(patch).toString();
  },
  selectCustomer: id =>
    knex('customer')
      .where('id', '=', id)
      .toString(),
  getCustomerByName: (cred, name) => {
    const {
      customers,
      // role
    } = cred.permissions;

    const getCustomerQuery = knex('customer').where('name', '=', name);

    if (cred.role !== 'admin') {
      getCustomerQuery.whereIn('id', customers);
    }

    return getCustomerQuery.toString();
  },
  selectCustomerIdByName: name =>
    knex('customer')
      .where('name', '=', name)
      .select('id')
      .toString(),
};

const Helpers = {
  getCustomerIdByName: async (sqlClient, name) => {
    const cidResult = await query(sqlClient, Sql.selectCustomerIdByName(name));

    const amount = R.length(cidResult);
    if (amount > 1) {
      throw new Error(
        `Multiple customer candidates for '${name}' (${amount} found). Do nothing.`,
      );
    }

    if (amount === 0) {
      throw new Error(`Not found: '${name}'`);
    }

    const cid = R.path(['0', 'id'], cidResult);

    return cid;
  },
};

const addCustomer = sqlClient => async (cred, input) => {
  if (cred.role !== 'admin') {
    throw new Error('Unauthorized.');
  }

  const prep = prepare(
    sqlClient,
    `CALL CreateCustomer(
        :id,
        :name,
        ${input.comment ? ':comment' : 'NULL'},
        ${input.private_key ? ':private_key' : 'NULL'}
      );
    `,
  );
  const rows = await query(sqlClient, prep(input));
  const customer = R.path([0, 0], rows);

  return customer;
};

const getCustomerByProjectId = sqlClient => async (cred, pid) => {
  const { customers, projects } = cred.permissions;
  const str = `
      SELECT
        c.id,
        c.name,
        c.comment,
        c.private_key,
        c.created
      FROM project p
      JOIN customer c ON p.customer = c.id
      WHERE p.id = :pid
      ${ifNotAdmin(
    cred.role,
    `AND (${inClauseOr([['c.id', customers], ['p.id', projects]])})`,
  )}
    `;
  const prep = prepare(sqlClient, str);

  const rows = await query(sqlClient, prep({ pid }));

  return rows ? rows[0] : null;
};

const deleteCustomer = sqlClient => async (cred, input) => {
  if (cred.role !== 'admin') {
    throw new Error('Unauthorized');
  }
  const prep = prepare(sqlClient, 'CALL deleteCustomer(:name)');

  const rows = await query(sqlClient, prep(input));

  // TODO: maybe check rows for changed values
  return 'success';
};

const getAllCustomers = sqlClient => async (cred, args) => {
  const where = whereAnd([
    args.createdAfter ? 'created >= :createdAfter' : '',
    ifNotAdmin(cred.role, `${inClause('id', cred.permissions.customers)}`),
  ]);
  const prep = prepare(sqlClient, `SELECT * FROM customer ${where}`);
  const rows = await query(sqlClient, prep(args));
  return rows;
};

const updateCustomer = sqlClient => async (cred, input) => {
  if (cred.role !== 'admin') {
    throw new Error('Unauthorized');
  }

  const { customers } = cred.permissions;
  const cid = input.id.toString();

  if (isPatchEmpty(input)) {
    throw new Error('input.patch requires at least 1 attribute');
  }

  await query(sqlClient, Sql.updateCustomer(cred, input));
  const rows = await query(sqlClient, Sql.selectCustomer(cid));

  return R.prop(0, rows);
};

const getCustomerByName = sqlClient => async (cred, args) => {
  const rows = await query(sqlClient, Sql.getCustomerByName(cred, args.name));
  return rows ? rows[0] : null;
};

const Queries = {
  addCustomer,
  deleteCustomer,
  getAllCustomers,
  getCustomerByProjectId,
  updateCustomer,
  getCustomerByName,
};

module.exports = {
  Sql,
  Queries,
  Helpers,
};
