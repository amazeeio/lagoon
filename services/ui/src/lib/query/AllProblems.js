import gql from 'graphql-tag';

export default gql`
  {
    allGroups {
      name
      projects {
        name
        id
        environments(type: PRODUCTION) {
          name
          id
          problems {
            id
            identifier
            severity
            source
          }
        }
      }
    }
  }
`;
