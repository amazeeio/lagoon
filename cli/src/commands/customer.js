// @flow

import { table } from 'table';
import { red } from 'chalk';
import R from 'ramda';

import { config } from '../config';
import gql from '../gql';
import { runGQLQuery } from '../query';
import {
  printProjectConfigurationError,
  printGraphQLErrors,
} from '../printErrors';

import typeof Yargs from 'yargs';
import type { BaseArgs } from '.';

export const command = 'customer';
export const description = 'Show customer details for a given project name';

export function builder(yargs: Yargs) {
  return yargs
    .usage(`$0 ${command} - ${description}`)
    .options({
      project: {
        demandOption: false,
        describe: 'Specify a project for the customer',
        type: 'string',
      },
    })
    .alias('p', 'project')
    .example(
      `$0 ${command}`,
      'Show customer details for the project configured in .lagoon.yml',
    )
    .example(
      `$0 ${command} -p myproject`,
      'Show customer details for the project "myproject"',
    );
}

type GetCustomerDetailsArgs = {
  project: string,
  clog: typeof console.log,
  cerr: typeof console.error,
};

export async function getCustomerDetails({
  project,
  clog,
  cerr,
}:
GetCustomerDetailsArgs): Promise<number> {
  const query = gql`
    query CustomerByProjectName($project: String!) {
      projectByName(name: $project) {
        customer {
          name
          comment
          private_key
          sshKeys {
            name
            keyValue
            keyType
            created
          }
          created
        }
      }
    }
  `;

  const result = await runGQLQuery({
    cerr,
    query,
    variables: { project },
  });

  const { errors } = result;
  if (errors != null) {
    return printGraphQLErrors(cerr, ...errors);
  }

  const customer = R.path(['data', 'projectByName', 'customer'], result);

  if (customer == null) {
    clog(red(`No customer found for project '${project}'!`));
    return 0;
  }

  const formatDeployPrivateKey = R.ifElse(
    R.identity,
    R.always('\u221A'),
    R.always('\u2717'),
  );
  const formatSshKeys = R.map(R.prop('name'));

  clog(`Customer details for project '${project}':`);
  clog(
    table([
      ['Name', R.prop('name', customer)],
      ['Comment', R.prop('comment', customer)],
      [
        'Deploy Private Key',
        formatDeployPrivateKey(R.prop('private_key', customer)),
      ],
      [
        'SSH Keys',
        R.join(', ', formatSshKeys(R.propOr([], 'sshKeys', customer))),
      ],
      ['Created', R.prop('created', customer)],
    ]),
  );

  return 0;
}

type Args = BaseArgs & {
  argv: {
    project: ?string,
  },
};

export async function handler({ argv, clog, cerr }: Args): Promise<number> {
  const project = R.prop('project', argv) || R.prop('project', config);

  if (project == null) {
    return printProjectConfigurationError(cerr);
  }

  return getCustomerDetails({ project, clog, cerr });
}
