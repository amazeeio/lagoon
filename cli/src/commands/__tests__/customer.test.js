// @flow

import { runGQLQuery } from '../../query';
import { getCustomerDetails } from '../customer';

jest.mock('../../query');

const _mock = (mockFn: any): JestMockFn => mockFn;

const mockResponse = {
  data: {
    projectByName: {
      customer: {
        name: 'customer1',
        comment: 'Comment about customer1',
        private_key: 'PRIVATE_KEY',
        sshKeys: [
          {
            name: 'a@example.com',
          },
        ],
        created: 'Wed May 18 2011 00:00:00 GMT+0000 (UTC)',
      },
    },
  },
};

describe('getCustomerDetails', () => {
  it('should show customer details', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve(mockResponse),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await getCustomerDetails({
      project: 'some_project',
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should show error message if GraphQL returns errors', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve({
        errors: [{ message: 'Something, something missing parameter X' }],
      }),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await getCustomerDetails({
      project: 'some_project',
      clog,
      cerr,
    });

    expect(code).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should show message for non-existing projects', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() => Promise.resolve({}));

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await getCustomerDetails({
      project: 'some_project',
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });
});
