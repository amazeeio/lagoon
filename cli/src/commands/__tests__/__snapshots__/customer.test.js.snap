// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCustomerDetails should show customer details 1`] = `
Array [
  Array [
    "Customer details for project 'some_project':",
  ],
  Array [
    "╔════════════════════╤═════════════════════════════════════════╗
║ Name               │ customer1                               ║
╟────────────────────┼─────────────────────────────────────────╢
║ Comment            │ Comment about customer1                 ║
╟────────────────────┼─────────────────────────────────────────╢
║ Deploy Private Key │ √                                       ║
╟────────────────────┼─────────────────────────────────────────╢
║ SSH Keys           │ a@example.com                           ║
╟────────────────────┼─────────────────────────────────────────╢
║ Created            │ Wed May 18 2011 00:00:00 GMT+0000 (UTC) ║
╚════════════════════╧═════════════════════════════════════════╝
",
  ],
]
`;

exports[`getCustomerDetails should show error message if GraphQL returns errors 1`] = `
Array [
  Array [
    "[31mOops! The lagoon API returned an error:[39m",
  ],
  Array [
    "[31m{[39m
[31m  \\"message\\": \\"Something, something missing parameter X\\"[39m
[31m}[39m",
  ],
]
`;

exports[`getCustomerDetails should show message for non-existing projects 1`] = `
Array [
  Array [
    "[31mNo customer found for project 'some_project'![39m",
  ],
]
`;
